crates/bevy_ecs/src/query/access.rs

```diff
@@ -423,6 +423,71 @@ impl<T: SparseSetIndex> FilteredAccess<T> {
     pub fn write_all(&mut self) {
         self.access.write_all();
     }
+
+    /// Returns the inverse of the filters contained in this access set.
+    pub fn invert_filters(&mut self) {
+        let filter_sets = self.filter_sets.drain(..);
+
+        #[derive(Debug, Copy, Clone)]
+        enum AccessType {
+            With(usize),
+            Without(usize),
+        }
+
+        // Gather and unify each filter set from separate bitsets for With and Without into
+        // a single collection to simplify logic
+        let mut accesses = vec![];
+        for filter in filter_sets {
+            let mut filters = vec![];
+            filters.extend(filter.with.ones().map(AccessType::With).collect::<Vec<_>>());
+            filters.extend(
+                filter
+                    .without
+                    .ones()
+                    .map(AccessType::Without)
+                    .collect::<Vec<_>>(),
+            );
+            accesses.push(filters);
+        }
+
+        let mut indices = vec![0; accesses.len()];
+        let lengths = accesses.iter().map(|v| v.len()).collect::<Vec<_>>();
+
+        loop {
+            let mut new_filter = AccessFilters::default();
+            for (el, index) in indices.iter().enumerate() {
+                match accesses[el][*index] {
+                    AccessType::With(id) => {
+                        new_filter.without.grow(id + 1);
+                        new_filter.without.insert(id);
+                    }
+                    AccessType::Without(id) => {
+                        new_filter.with.grow(id + 1);
+                        new_filter.with.insert(id);
+                    }
+                }
+            }
+            self.filter_sets.push(new_filter);
+
+            let mut update_index = indices.len() - 1;
+            loop {
+                // update while rolling over indices
+                indices[update_index] = (indices[update_index] + 1) % lengths[update_index];
+
+                // if index did not roll over, or we are at the first element, stop looping
+                if indices[update_index] != 0 || update_index == 0 {
+                    break;
+                }
+
+                update_index -= 1;
+            }
+
+            // if we have rolled over the first element, we have gone through all combinations
+            if update_index == 0 && indices[0] == 0 {
+                break;
+            }
+        }
+    }
```

Comment:
This feels way more complicated than what I had imagined. I was thinking of something like this:

```rs
    pub fn invert_filters(&mut self) {
        let mut acc = FilteredAccess::default();

        for filter_set in self.filter_sets.drain(..) {
            let mut new = FilteredAccess::default();
            for with in filter_set.with.ones() {
                let mut tmp = FilteredAccess::default();
                tmp.and_without(with);
                new.append_or(&tmp);
            }
            for without in filter_set.with.ones() {
                let mut tmp = FilteredAccess::default();
                tmp.and_with(without);
                new.append_or(&tmp);
            }
            acc.extend(&new);
        }

        *self = acc;
    }
```

The idea is that `!((... & ...) | (... & ...) | ...) = !(... & ...) & !(... & ...) & ... = (!... | !...) & (!... | !...) & ...`, so you just need to invert the singlular terms in each conjunction (using `and_without` for the existing `with` and `and_with` for the existing `without`), combine them in a disjunction (using `append_or`, but could be done more efficiently) and then combine everything together in a conjunction (using `extend`).

---

I'm also not sure what should be done with `self.accesses` (not to be confused with your `accessed` local variable)